app/controllers/devise/two_factor_authentication_controller.rb

```diff
@@ -49,14 +56,20 @@ def update
 
     private
 
-    def check_already_authenticated
-      redirect_to profile_path if user_fully_authenticated?
+    def valid_otp?(code)
+      case current_otp_method
+      when :totp
+        return resource.authenticate_totp(code)
+      when :rescue_code
+        code.tr!('-', '')
+        return resource.authenticate_backup_code(code)
+      else
+        return resource.authenticate_direct_otp(code)
+      end
```

Comment:
Why 3 cases? Doesn't the two_factor_authentication gem already handle the first and last cases? We only called `authenticate_otp` before, so why would the addition of backup codes require us to call an additional 2 methods? 

Also, why not process rescue codes in their own controller? It would separate the concerns and stop perpetuating the use of params, which is not a good design choice for this problem IMO. Once you find yourself writing case statements like this, or other similar conditionals based on the type or kind of something, that's usually an indication of a code smell.